src/sinks/datadog/traces/request_builder.rs

```diff
@@ -0,0 +1,351 @@
+                            };
+                            let mut compressor = Compressor::from(self.compression);
+                            match compressor.write_all(&payload) {
+                                Ok(()) => {
+                                    results.push(Ok((metadata, compressor.into_inner().freeze())))
+                                }
+                                Err(_) => results.push(Err(RequestBuilderError::FailedToEncode {
+                                    reason: "Payload compression failed.",
+                                    dropped_events: n as u64,
+                                })),
+                            }
+                        }
+                        Err(err) => results.push(Err(RequestBuilderError::FailedToEncode {
+                            reason: err.parts().0,
+                            dropped_events: err.parts().1,
+                        })),
+                    })
+            }
+        }
+        results
+    }
+
+    fn build_request(&mut self, metadata: Self::Metadata, payload: Self::Payload) -> Self::Request {
+        let mut headers = BTreeMap::<String, String>::new();
+        headers.insert(
+            "Content-Type".to_string(),
+            "application/x-protobuf".to_string(),
+        );
+        headers.insert("DD-API-KEY".to_string(), metadata.api_key.to_string());
+        headers.insert(
+            "X-Datadog-Reported-Languages".to_string(),
+            metadata.lang.unwrap_or_else(|| "".into()),
+        );
+        if let Some(ce) = self.compression.content_encoding() {
+            headers.insert("Content-Encoding".to_string(), ce.to_string());
+        }
+        TraceApiRequest {
+            batch_size: metadata.batch_size,
+            body: payload,
+            headers,
+            finalizers: metadata.finalizers,
+            uri: self
+                .endpoint_configuration
+                .get_uri_for_endpoint(metadata.endpoint),
+        }
+    }
+}
+
+pub struct DatadogTracesEncoder {
+    max_size: usize,
+}
+
+#[derive(Debug, Snafu)]
+pub enum EncoderError {
+    #[snafu(display("Unable to split payload into small enough chunks"))]
+    UnableToSplit { dropped_events: u64 },
+}
+
+impl EncoderError {
+    pub const fn parts(&self) -> (&'static str, u64) {
+        match self {
+            Self::UnableToSplit { dropped_events: n } => ("unable to split into small chunks", *n),
+        }
+    }
+}
+impl DatadogTracesEncoder {
+    fn encode_trace(
+        &self,
+        key: &PartitionKey,
+        events: Vec<TraceEvent>,
+    ) -> Vec<Result<(Vec<u8>, Vec<TraceEvent>), EncoderError>> {
+        let mut encoded_payloads = Vec::new();
+        let payload = DatadogTracesEncoder::trace_into_payload(key, &events);
+        let encoded_payload = payload.encode_to_vec();
+        // This may happen exceptionally
+        if encoded_payload.len() > self.max_size {
+            emit!(DatadogTracesPayloadTooBig {});
```

Comment:
Ah, I see, I think I misunderstood the behavior here initially. It seems like the sink handles this situation fine by breaking up the batches? If so, I'd actually suggest dropping this internal event. We could add a debug message. I think users typically won't care about this.

Reply:
I think tracking that can be interesting as this should not happen, but if this happens it might show that processed traces are really big (which may happen if tracing is enabled at the very top-level of a large application with some non-trace-friendly code with recursion). Dropped traces will always be counted, so letting a debug message mays be enough for troubleshooting if needed.

Reply:
I agree, it will happen if we have very large traces incoming, anyway dropped traces will be counted, if a problem needs to be investigated, adding a debug log here is probably enough.